### PR TITLE
Adjust terminal tests to new behavior in JDK 22.

### LINKED ISSUE #98033: Adjust terminal tests to jdk22 behavior
In JDK22, the behavior of System.console() when the terminal is redirected will change. System.console() will no longer return null in this case. Our Terminal abstraction currently relies on seeing null returned to decide which implementation to use. This issue is a reminder to adjust our Terminal code to understand the new jdk behavior.

See https://bugs.openjdk.org/browse/JDK-8308591

### PR DESCRIPTION
**Pull request by mosche**
JDK 22 may return a console even if the terminal is redirected. These cases are detected using the new `Console#isTerminal()` to maintain the current behavior.

Closes #98033

### CODE REVIEW

**Comment by mosche** [comment]
Wondering, I would have expected `SystemTerminal` to overwrite `isHeadless`, but that's not the case. `isHeadless` always returns `false`. Is that a bug or just some left over?

**Comment by mosche** [comment]
Not sure how much value there is in having that new test case. 
Though, it does validates the change if running on Java 22. WDYT?

**Review by rjernst — COMMENTED** [review]
This seems fine, but perhaps we should add the system property `-Djdk.console=java.base` to SystemJvmOptions, then we won't need to change the Terminal detection for now?

**Comment by mosche** [comment]
Sure, I'll just do that 👍

**Review by mark-vieira — APPROVED** [review]

**Comment by mosche** [comment]
@rjernst apparently there's more to it. Other than release notes in https://bugs.openjdk.org/browse/JDK-8309155 suggest, `-Djdk.console=java.base` doesn't completely restore the previous behavior. If added to `SystemJvmOptions`, I'm reading `null` for `bootstrap.password` on startup from the terminal (on JDK 22) causing an NPE:

```
Exception in thread "main" java.lang.NullPointerException: Cannot read the array length because "array" is null
        at java.base/java.nio.CharBuffer.wrap(CharBuffer.java:468)
        at org.elasticsearch.common.settings.KeyStoreWrapper.setString(KeyStoreWrapper.java:586)
        at org.elasticsearch.cli.keystore.AddStringKeyStoreCommand.executeCommand(AddStringKeyStoreCommand.java:99)
```

I suppose as far as the terminal checks are concerned, we should rely on `isTerminal()` / the current version of the code here as this works seamlessly.

Let me know if you think we need to better understand the behavior (change) if using `-Djdk.console=java.base` and I'll make sure to investigate that further.

File: libs/cli/src/main/java/org/elasticsearch/cli/Terminal.java [hunk 1]
```diff
@@ -285,6 +287,23 @@ static boolean isSupported() {
...
+            // https://bugs.openjdk.org/browse/JDK-8308591
+            Console console = System.console();
+            if (console != null && Runtime.version().feature() >= JDK_VERSION_WITH_IS_TERMINAL) {
+                try {
+                    // verify the console is a terminal using isTerminal() on JDK >= 22
+                    // TODO: Remove reflection once Java 22 sources are supported, e.g. using a MRJAR
+                    Method isTerminal = Console.class.getMethod("isTerminal");
+                    return Boolean.TRUE.equals(isTerminal.invoke(console)) ? console : null;
+                } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+                    throw new RuntimeException(e);
```
Comment (rjernst) [inline libs/cli/src/main/java/org/elasticsearch/cli/Terminal.java:301]: nit: I would make this AssertionError, if we hit this, something is very wrong (jdk22 is missing a method).

File: libs/cli/build.gradle [hunk 2]
```diff
@@ -11,9 +11,18 @@ apply plugin: 'elasticsearch.publish'
...
+
+  testImplementation(project(":test:framework")) {
+    exclude group: 'org.elasticsearch', module: 'elasticsearch-cli'
+  }
+}
+
+tasks.named("test").configure {
+  // Disable security to force new behavior in JDK 22.
+  // If enabled, JDK 22 doesn't provide a console if redirected.
+  systemProperty 'tests.security.manager', 'false'
```
Comment (rjernst) [inline libs/cli/build.gradle:23]: We should be able to avoid a blanket disabling by adding `@WithoutSecurityManager` to the new test class.

**Review by rjernst — APPROVED** [review]
LGTM

File: libs/cli/build.gradle (hunk 2 above)
Comment (mosche) [reply libs/cli/build.gradle:23]: Great, thanks!

**Comment by mosche** [comment]
@elasticmachine update branch

**Pull request merged**